django/curator/tests/test_tag_deduplication.py

```diff
@@ -0,0 +1,117 @@
+import random
+import shortuuid
+import string
+
+from django.test import TestCase
+from taggit.models import Tag
+
+from curator.tag_deduplication import TagPreprocessing, TagClustering, TagGazetteering
+from curator.models import CanonicalTag, CanonicalTagMapping
+
+
+class TagPreprocessingTestCase(TestCase):
+    def setUp(self):
+        pass
+
+    # Make sure tag preprocessing
+    def test_split_by_and(self):
+        preprocessed_Tags = TagPreprocessing.preprocess("biology and chemistry")
+        self.assertEqual(
+            ["biology", "chemistry"], preprocessed_Tags, "Failed to split by 'and'"
+        )
+        self.assertEqual(list, type(preprocessed_Tags))
+
+    def test_split_by_semicolon(self):
+        preprocessed_Tags = TagPreprocessing.preprocess("agent-based model; land-use")
+        self.assertEqual(
+            ["agent-based model", "land-use"],
+            preprocessed_Tags,
+            "Failed to split by ';'",
+        )
+        self.assertEqual(list, type(preprocessed_Tags))
+
+    def test_split_by_comma(self):
+        preprocessed_Tags = TagPreprocessing.preprocess("agent-based model, land-use")
+        self.assertEqual(
+            ["agent-based model", "land-use"],
+            preprocessed_Tags,
+            "Failed to split by ','",
+        )
+        self.assertEqual(list, type(preprocessed_Tags))
+
+
+class TestTagClustering(TestCase):
+    def setUp(self):
+        for i in range(300):
+            id = TestTagClustering._id_generator()
+            CanonicalTag(name=id).save()
+            Tag(name=id + "1").save()
+            Tag(name=id + "2").save()
+            Tag(name=id + "3").save()
+
+    def test_uncertain_pairs(self):
+        tag_clustering = TagClustering()
+        uncertain_pairs = tag_clustering.uncertain_pairs()
+        self.assertEqual(list, type(uncertain_pairs))
+        self.assertEqual(1, len(uncertain_pairs))
+        self.assertEqual(2, len(uncertain_pairs[0]))
+        self.assertEqual(int, type(uncertain_pairs[0][0]["id"]))
+        self.assertEqual(str, type(uncertain_pairs[0][0]["name"]))
+        self.assertEqual(str, type(uncertain_pairs[0][0]["slug"]))
```

Comment:
are there any assertions you can make about the contents of these data structures in addition to their types and shapes?